File: server/php/src/Presentation/Middleware/GlobalExceptionMiddleware.php
<?php

namespace OnlinePayments\ExampleApp\Presentation\Middleware;

use Fig\Http\Message\StatusCodeInterface;
use OnlinePayments\ExampleApp\Application\Exceptions\SdkException;
use OnlinePayments\ExampleApp\Presentation\Models\Errors\SystemErrorResponse;
use OnlinePayments\ExampleApp\Presentation\Models\Errors\ValidationErrorResponse;
use OnlinePayments\ExampleApp\Presentation\Extensions\ResponseExtension;
use OnlinePayments\ExampleApp\Configuration\Exceptions\ValidationException;
use Psr\Http\Server\RequestHandlerInterface;
use Psr\Log\LoggerInterface;
use Slim\Psr7\Request as SlimRequest;
use Slim\Psr7\Response as SlimResponse;
use InvalidArgumentException;

final readonly class GlobalExceptionMiddleware
{
    private LoggerInterface $logger;

    public function __construct(LoggerInterface $logger)
    {
        $this->logger = $logger;
    }

    public function __invoke(SlimRequest $request, RequestHandlerInterface $handler): SlimResponse
    {
        try {
            $responseFromHandler = $handler->handle($request);

            return ResponseExtension::toSlim($responseFromHandler);
        } catch (\Throwable $ex) {
            $this->logger->error('Exception occurred while processing request.', ['exception' => $ex]);

            [$statusCode, $errorResponse] = $this->mapExceptionToResponse($ex);

            $body = json_encode($errorResponse, JSON_UNESCAPED_SLASHES | JSON_UNESCAPED_UNICODE);

            $response = new SlimResponse();
            $response->getBody()->write($body);

            return $response
                ->withStatus($statusCode)
                ->withHeader('Content-Type', 'application/problem+json');
        }
    }

    private function mapExceptionToResponse(\Throwable $ex): array
    {
        if ($ex instanceof SdkException) {
            return [
                $ex->getStatusCode(),
                new ValidationErrorResponse($ex->getMessage())
            ];
        }

        if ($ex instanceof InvalidArgumentException) {
            return [
                StatusCodeInterface::STATUS_BAD_REQUEST,
                new ValidationErrorResponse($ex->getMessage())
            ];
        }

        if ($ex instanceof ValidationException) {
            return [
                StatusCodeInterface::STATUS_BAD_REQUEST,
                new ValidationErrorResponse($ex->errors[0] ?? $ex->getMessage())
            ];
        }

        return [
            StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR,
            new SystemErrorResponse(
                StatusCodeInterface::STATUS_INTERNAL_SERVER_ERROR,
                'Internal Server Error.'
            )
        ];
    }
}

File: server/ruby/app/presentation/middleware/global_exception_middleware.rb
require 'json'

module Presentation
  module Middleware
    class GlobalExceptionMiddleware
      def initialize(app, logger = Rails.logger)
        @app = app
        @logger = logger
      end

      def call(env)
        @app.call(env)
      rescue => ex
        log_exception(ex)
        status, error_response = map_exception_to_response(ex)

        body = error_response.to_json
        headers = { 'Content-Type' => 'application/problem+json' }

        [status, headers, [body]]
      end

      private

      def map_exception_to_response(ex)
        case ex
        when Business::Exceptions::ValidationErrorResponse
          [400, { message: ex.message.to_s }]
        when Business::Exceptions::SdkException
          [ex.status_code || 400, { message: ex.message.to_s }]
        when ArgumentError
          [400, { message: ex.message.to_s }]
        else
          [500, { message: 'Internal Server Error.' }]
        end
      end

      private

      def extract_sdk_error_message(ex)
        if ex.respond_to?(:errors) && ex.errors.any?
          first_error = ex.errors.first
          first_error.message.presence ||
            first_error.id.presence ||
            first_error.to_s
        elsif ex.respond_to?(:api_message)
          ex.api_message.to_s
        else
          ex.message.to_s
        end
      rescue
        "An unexpected error occurred."
      end

      def log_exception(ex)
        @logger.error("Exception occurred while processing request: #{ex.class} - #{ex.message}")
        @logger.error(ex.backtrace.join("\n")) if ex.backtrace
      end
    end
  end
end